espnet2/iterators/sequence_iter_factory.py

```diff
@@ -140,6 +143,18 @@ def build_iter(self, epoch: int, shuffle: bool = None) -> DataLoader:
         else:
             kwargs = {}
 
+        # reshuffle whole 'batches' so that elements within a batch can move
+        # between different batches
+        if self.shuffle_within_batch:
```

Comment:
Will it change the result?
If we are using the basic loss, the pooling operation across the sentence is order-agnostic, and it would not change the result (except for the numerical difference).

Reply:
Let's try to pass this part by a test.

Reply:
For speaker recognition, having diverse speaker classes within each mini-batch is critical. 
I looked at many ESPnet dataset code, but found that `batch_sampler` in ESPnet only plays with the order of minibatchs between epochs but don't randomly shuffle the samples that goes into each mini-batch.

I believe this is related to the nature of ASR and some other tasks where we sort in length. 
In contrast, in speaker recognition, we need same length utterances within the mini-batch and also needs to belong to different speakers. 

I empirically found that without `shuffle_within_batch`, even the training loss doesn't decrease at all.